pennylane_lightning/lightning_gpu/lightning_gpu.py

```diff
@@ -0,0 +1,754 @@
+            if self.shots is not None:
+                warn(
+                    "Requested adjoint differentiation to be computed with finite shots."
+                    " The derivative is always exact when using the adjoint differentiation method.",
+                    UserWarning,
+                )
+
+            tape_return_type = self._check_adjdiff_supported_measurements(measurements)
+
+            if math.allclose(grad_vec, 0) or tape_return_type is None:
+                return lambda tape: math.convert_like(
+                    np.zeros(len(tape.trainable_params)), grad_vec
+                )
+
+            if tape_return_type is Expectation:
+                if len(grad_vec) != len(measurements):
+                    raise ValueError(
+                        "Number of observables in the tape must be the same as the length of grad_vec in the vjp method"
+                    )
+
+                if np.iscomplexobj(grad_vec):
+                    raise ValueError(
+                        "The vjp method only works with a real-valued grad_vec when the tape is returning an expectation value"
+                    )
+
+                ham = qml.Hamiltonian(grad_vec, [m.obs for m in measurements])
+
+                # pylint: disable=protected-access
+                def processing_fn(tape):
+                    nonlocal ham
+                    num_params = len(tape.trainable_params)
+
+                    if num_params == 0:
+                        return np.array([], dtype=self.state.dtype)
+
+                    new_tape = tape.copy()
+                    new_tape._measurements = [qml.expval(ham)]
+
+                    return self.adjoint_jacobian(new_tape, starting_state, use_device_state)
+
+                return processing_fn
+
+        # pylint: disable=attribute-defined-outside-init
+        def sample(self, observable, shot_range=None, bin_size=None, counts=False):
+            if observable.name != "PauliZ":
+                self.apply_lightning(observable.diagonalizing_gates())
+                self._samples = self.generate_samples()
+            return super().sample(
+                observable, shot_range=shot_range, bin_size=bin_size, counts=counts
+            )
+
+        # pylint: disable=missing-function-docstring
+        def generate_samples(self):
+            """Generate samples
+
+            Returns:
+                array[int]: array of samples in binary representation with shape
+                ``(dev.shots, dev.num_wires)``
+            """
+            return self.measurements.generate_samples(len(self.wires), self.shots).astype(
+                int, copy=False
+            )
+
+        # pylint: disable=protected-access, missing-function-docstring
+        def expval(self, observable, shot_range=None, bin_size=None):
+            if observable.name in [
+                "Projector",
+            ]:
+                return super().expval(observable, shot_range=shot_range, bin_size=bin_size)
```

Comment:
I guess we never use Projectors in the tests to hit this. Can we add a simple test?

Reply:
yea, we will never hit is since it's not an allowed_obs in `lightning.gpu`. I'm going to remove it.